### ISSUE TITLE
Improve sidecar mode for Kubernetes

### ISSUE THREAD

**Issue by jsanda** [thelastpickle/cassandra-reaper#956]
# Improve sidecar mode for Kubernetes
The [docs](http://cassandra-reaper.io/docs/usage/sidecar_mode/) states that sidecar mode is _Kubernetes Friendly_. The purpose of this ticket is to highlight and address some problems with Reaper sidecar in Kubernetes.

I added initial support for Reaper sidecar mode in [Cass Operator](https://github.com/datastax/cass-operator), a Kubernetes operator for Cassandra. The initial integration does not work when Cassandra authentication is enabled (see this [issue](https://github.com/datastax/cass-operator/issues/211) for details). While working on the fix, I have encountered problems in the following areas:

* health probe
* startup initialization code
* schema updates

In Kubernetes, Cassandra is deployed in a StatefulSet. A StatefulSet consists of one or more pods. A pod can have multiple application containers. When Cass Operator deploys Reaper, each Cassandra pod includes a `cassandra` container as well as a `reaper` container. Containers can have readiness and liveness probes. Kubernetes uses the readiness probe to determine when a container is ready to start serving requests. It uses the liveness probe to determine whether the container needs to be restarted (see [here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) for more details about Kubernetes probes).

I am using Reaper's `healthcheck` endpoint for both the liveness and readiness probes. This endpoint is handled by the class `ReaperHealthCheck`. `ReaperHealthCheck` first checks that it is connected to the backend storage and then queries it.

Here is where things get interesting. The startup order of application containers in a pod is effectively non-deterministic. Cassandra can start before Reaper and vice versa. A pod is considered ready when all of its application containers report ready. For the Cassandra pod this means that both the `cassandra` and the `reaper` containers need to report ready in order for the pod to be considered ready. 

Cass Operator create roles in Cassandra, including one for Reaper, but they do not get created until the C* cluster is ready. The `reaper` container cannot reach the ready state because it does not have credentials with which to connect to Cassandra. And the pods cannot reach the ready state since the `reaper` containers cannot reach the ready state.

Initially I updated Cass Operator so that it doe the following:

* Deploy Cassandra without Reaper
* Wait for Cassandra to be ready
* Create the roles
* Update StatefulSet definition to include Reaper

From a user perspective this looks like a cluster-wide rolling restart. In my testing thus far I found that this approach works, but it is suboptimal. It really slows down the overall deployment time. I want to deploy Reaper at the same time as Cassandra and avoid having to do a rolling restart. 

There is another issue with which I have been struggling around appyling schema changes. The schema migration library that Reaper uses implements coordination via table-based locks. The locks are based on IP addresses. The readiness and/or liveness probes frequently fail before all schema changes are applied. Kubernetes restarts the container, and because it is assigned a different IP address, the `reaper` container cannot immediately reacquire the lock. This results in an ugly cycle of container restarts before things eventually stabalize.

I want to make the following changes to address these problems:

* Change `ReaperHealthCheck` such that it only checks if Reaper is running and do not check storage connectivity
* Change startup/initialization code so that it does not block waiting to connect to Cassandra
* Make it easier to apply schema changes

Note that the changes I am proposing are only needed for the sidecar scenario in Kubernetes. If Reaper is deployed in a different mode, these issues can be avoided.

Reaper will be reporting ready/healthy even before it is fully initialized. This will help avoid k8s liveness and readiness probe failures that can occur while Reaper waits for Cassandra to come up. In order for the `/healthcheck` endpoint to be available though, startup cannot block waiting to establish a connection to the storage backend. Some initialization will have to be done in a background thread.

If a client invokes an endpoint before Reaper has comepleted initialization tasks, Reaper should fail gracefully. We can use a JAX-RS filter to intercept all endpoints and return an appropriate error in this situation.

We need to make it easier to apply schema changes. To be honest, I don't think that the coordination provided by the schema migration library is very useful. We have to first create the `reaper_db` keyspace before we can apply any schema changes. We need some form of coordination for this. In Kubernetes, I create the keyspace with a Job. There is guaranteed to be only a single instance of the Job running, which means no lock contetion. I would like to reuse and extend that Job to also apply the schema changes. I am currently doing this with a modified version of Reaper that exits after applying the schema changes. This is a rather heavyweight solution though. I think we really just need to run `CassandraStorage.initializeAndUpgradeSchema()`.

I am propsing these changes within the context of the work I am doing for Cass Operator. There are several other Cassandra operator projects. It is worth pointing out that these changes are not specific to Cass Operator. They will benefit other operators that also want to run Reaper in sidecar mode.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-454) by [Unito](https://www.unito.io)
┆Issue Number: K8SSAND-454

**Comment by jsanda** [comment]
In my initial write up, I mentioned making schema changes easier. Right now in #957 I simply check for a flag being set like this in `ReaperApplication`:

```java
if (Boolean.parseBoolean(System.getenv("SCHEMA_ONLY"))) {
  return;
}
```

This works but it is a heavyweight solution. 

A better approach would be to move the `io.cassandrareaper.storage` and subpackages under a new set of maven modules at `src/storage`. It would look something like this:

```
src/storage/
├── cassandra
├── cassandra-schema-installer
├── memory
└── postgres
```

The `cassandra`, `memory`, and `postgres` modules would contain existing code. The main `server` module would add these add dependencies. The `cassandra-schema-installer` module would depend only on the `cassandra` module and produce an executable JAR as its artifact. This JAR would lend itself much better to the schema installation job.

Reorganizing a project like this can be tedious and messy. While I think that these changes would be good, I would like to get some feedback before I pursue them.

@adejanovski and/or @michaelsembwever any thoughts?

**Comment by michaelsembwever** [comment]
Is it easy today to run the Cass Operator with Reaper in LOCAL/EACH mode instead of SIDECAR mode?

Using either LOCAL/EACH mode seems the correct solutio here. 

You want Reaper's healthcheck to be "true". You need the C* pods to be running first. You want to be able to run one Reaper pod, let it initialise, before scaling out further Reaper pods. 

Given this, why do we want to hack Reaper (in a number of ways) to make it work with sidecar mode? What's the advantages to that, when it encourages a design that uses more resources? The value of sidecar mode is to avoid having to open C* JMX ports, but inside a k8 cluster that's not an issue (because it is easy to open the jmx port to just the reaper pod).

**Comment by jsanda** [comment]
Thanks @michaelsembwever for taking the time to review and reply :)

>Is it easy today to run the Cass Operator with Reaper in LOCAL/EACH mode instead of SIDECAR mode?

It is not implemented yet, but it will be easier to support. The problems described in this issue for the most part are specific to sidecar mode.

>You want Reaper's healthcheck to be "true". You need the C* pods to be running first.

I changed this behavior with my PR #957. The code looks like this:

```java
 protected Result check() throws ReaperException {
    if (context.config.isInKubernetesSidecarMode()) {
      // If Reaper is running in Kubernetes as a sidecar, i.e., in the same pod as the
      // cassandra container, then we do not want the health check to depend on Cassandra.
      // See https://github.com/thelastpickle/cassandra-reaper/issues/956 for details.
      if (context.isRunning.get()) {
        return Result.healthy();
      } else {
        return Result.unhealthy("Reaper is not running");
      }
    }
    // existing, non-k8s behavior follows...
  }
```
>You want to be able to run one Reaper pod, let it initialise, before scaling out further Reaper pods.

Are you referring to sidecar mode in particular?

>Given this, why do we want to hack Reaper (in a number of ways) to make it work with sidecar mode? What's the advantages to that, when it encourages a design that uses more resources?

As I mentioned in the description of this issue, the docs state that sidecar mode is _Kubernetes Friendly_. This is not true. I do not consider the changes hacks. I put a lot of thought and testing into them already. I also discussed them briefly with @adejanovski before creating this ticket. With that said, unless I am unaware of some compelling reasons for sidecar mode, I would not recommend it as a first option in Kubernetes. 

>The value of sidecar mode is to avoid having to open C* JMX ports, but inside a k8 cluster that's not an issue (because it is easy to open the jmx port to just the reaper pod).

The process for exposing a port in k8s is different, but doing so does still carry security concerns.

**Comment by michaelsembwever** [comment]
> > You want Reaper's healthcheck to be "true". You need the C* pods to be running first.

> I changed this behavior with my PR #957. The code looks like this:
> `if (context.config.isInKubernetesSidecarMode()) {`

I'm not in favour of this approach. As mentioned, this is a hack to work-around the side-car mode not being the appropriate design choice for kubernetes.

I would… 
a) implement Cass Operator to use either LOCAL/EACH, and
b) implement a dropwizard `db` command, instead of the `"SCHEMA_ONLY"` flag approach in #957 

For (b), references are 
- https://www.dropwizard.io/en/latest/getting-started.html#running-your-application
- https://github.com/dropwizard/dropwizard/blob/master/dropwizard-core/src/main/java/io/dropwizard/Application.java#L102-L105

> > You want to be able to run one Reaper pod, let it initialise, before scaling out further Reaper pods.

> Are you referring to sidecar mode in particular?

No, this is generally the safe approach to installing the schema for the first time, regardless of mode.

> As I mentioned in the description of this issue, the docs state that sidecar mode is Kubernetes Friendly. This is not true. 

Correct. That statement should be removed from the docs.

>  I do not consider the changes hacks. … With that said, unless I am unaware of some compelling reasons for sidecar mode, I would not recommend it as a first option in Kubernetes.

I'm calling it a "hack" because you are having to introducing a modified sidecar mode to get it to work with kubernetes. Or as I expressed it above, sidecar is _obviously_ the wrong design choice for Kubernetes, so forcing it is a hack IMHO.

**Comment by jsanda** [comment]
To the best of my knowledge sidecar mode was not built with k8s in mind and there have been some challenges as a result. I do not want to be so quick though to write it off as the wrong choice for k8s on that basis on alone. I think the decision should be made based on the merits of sidecar mode. If there are enough benefits of sidecar mode that are also applicable in k8s, then I am happy push to make sidecar mode more k8s friendly. If that is not the case though, then I will happily be on my way :)

I want to clarify an earlier statement. I said that I would not recommend sidecar mode as a first choice for k8s. This is primarily due to the additional resource requirements of running a Reaper instance per C* node. Reaper could be more modularized to help reduce those requirements. For example, the thread pools for running various scheduled jobs could be replaced by k8s jobs. The jobs could be lightweight and simply call the appropriate Reaper endpoint to kick off necessary work. Or better yet, the job could include the necessary work, making Reaper more modular, more distributed, more fault tolerant, etc.

**Comment by michaelsembwever** [comment]
>Is it easy today to run the Cass Operator with Reaper in LOCAL/EACH mode instead of SIDECAR mode?

Please can we have cass-operator support LOCAL/EACH mode. 
I would (and will) not be recommending SIDECAR mode in anyone's k8s setup. I cannot justify the cost. It's just not possible to explain to people to run an additional 88 JVMs (each 2CPU and 2GB ram) on a 100 node cluster. The cost of that is truly astounding, and adds zero value. Two reaper instances per data-center provides a high-availability full-value (all features!) setup.

Even with a more modularised Reaper, the additional resource costs of all those sidecar pods, when all you need is one or two reaper pods per data-center, cannot be justified. The notion of heavy JVM co-processes inside k8s pods doesn't seem to be best practices either.

**Comment by markharding** [comment]
Since the 2.2 update the sidecars are unable to start:

```
INFO   [2021-02-22 13:29:07,086] [main] o.c.c.m.MigrationRepository - Found 12 migration scripts 
WARN   [2021-02-22 13:29:07,086] [main] i.c.s.CassandraStorage - Starting db migration from 25 to 27… 
ERROR  [2021-02-22 13:29:08,328] [main] i.c.ReaperApplication - Storage is not ready yet, trying again to connect shortly... 
java.lang.IllegalStateException: Database migration can not happen with other reaper instances running. Found  [[10.156.4.3, 127.0.0.1, 127.0.0.1, 127.0.0.1, 10.156.1.2]]
	at com.google.common.base.Preconditions.checkState(Preconditions.java:534)
	at io.cassandrareaper.storage.CassandraStorage.initializeCassandraSchema(CassandraStorage.java:288)
	at io.cassandrareaper.storage.CassandraStorage.initializeAndUpgradeSchema(CassandraStorage.java:253)
	at io.cassandrareaper.storage.CassandraStorage.<init>(CassandraStorage.java:241)
	at io.cassandrareaper.ReaperApplication.initializeStorage(ReaperApplication.java:480)
	at io.cassandrareaper.ReaperApplication.tryInitializeStorage(ReaperApplication.java:303)
	at io.cassandrareaper.ReaperApplication.run(ReaperApplication.java:181)
	at io.cassandrareaper.ReaperApplication.run(ReaperApplication.java:98)
	at io.dropwizard.cli.EnvironmentCommand.run(EnvironmentCommand.java:43)
	at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:87)
	at io.dropwizard.cli.Cli.run(Cli.java:78)
	at io.dropwizard.Application.run(Application.java:93)
	at io.cassandrareaper.ReaperApplication.main(ReaperApplication.java:117)
```

Is there a quick workaround?

**Comment by adejanovski** [comment]
yes, it's an upgrade "feature" 😅 
Truncate the `running_reapers` table and then you'll be good to go.

**Comment by adejanovski** [comment]
@jsanda, I assume we can close this issue now. Re-open it if you think there's still something to address here.
